### DISCUSSION TITLE
Prevent V+ config overwrite

### DISCUSSION THREAD

**Discussion by nachtgold** [mbround18/valheim-docker#341] [Help!] [answered]
Hello, 

I setup a valheim server some weeks ago. Now we want to use some tweaks and updated to lastest dockerfile and set "TYPE: ValheimPlus". But if I change values in `server/BepInEx/config/valheim_plus.cfg` and restart the server, my changes are gone. 

I looked into #308 and the undocumented MODS_LOCATION option, but none of them helped me, to find out what is the problem. 

BR Kai

**Comment by mbround18** [comment]
Have you pulled a latest image recently? This was a bug before when it came to extracting but we patched that up

**Comment by nachtgold** [comment, marked as answer]
I know, therefore I compared the hashes of the used image with yours from https://hub.docker.com/layers/mbround18/valheim/latest/images/sha256-65a2277f7a6d415d48a43ceb4044671a47843505d49fe34434e230b3919c64a4?context=explore and they match. Can I check the code inside the container?

~~PS: how can I activate the debug loggings? Maybe they can help "debugging". Can the parameter DEBUG_MODE (https://github.com/mbround18/valheim-docker/blob/62249670b33e3c5459a9dea2bee0a058e64450d8/src/odin/README.md) of odin passed into the image?~~

Yes it does.

```
> less valheim_server.log | grep DEBUG
[ODIN][DEBUG] - Env var default 'LD_LIBRARY_PATH': '/home/steam/valheim/linux64'
[ODIN][DEBUG] - Setting up base command
[ODIN][DEBUG] - Env var default 'SERVER_EXTRA_LAUNCH_ARGS': '-nographics -batchmode'
[ODIN][DEBUG] - Env var default 'APPID': '892970'
[ODIN][DEBUG] - Env var found 'TYPE': 'valheimplus'
[ODIN][DEBUG] - Password found, adding password flag.
[ODIN][DEBUG] - Parsing Doorstop locations.
[ODIN][DEBUG] - Env var default 'DOORSTOP_LIB': 'libdoorstop_x64.so'
[ODIN][DEBUG] - Env var default 'DOORSTOP_LIBS': '/home/steam/valheim/doorstop_libs'
[ODIN][DEBUG] - Path /home/steam/valheim/doorstop_libs exists
[ODIN][DEBUG] - Env var default 'DOORSTOP_INVOKE_DLL_PATH': '/home/steam/valheim/BepInEx/core/BepInEx.Preloader.dll'
[ODIN][DEBUG] - Env var default 'DOORSTOP_CORLIB_OVERRIDE_PATH': '/home/steam/valheim/unstripped_corlib'
[ODIN][DEBUG] - Path /home/steam/valheim/unstripped_corlib exists
[ODIN][DEBUG] - Parsing LD locations.
[ODIN][DEBUG] - Env var default 'LD_PRELOAD': ''
[ODIN][DEBUG] - Env var default 'LD_LIBRARY_PATH': './linux64:/home/steam/valheim/doorstop_libs'
[ODIN][DEBUG] - Parsing LD locations.
[ODIN][DEBUG] - Env var default 'DYLD_LIBRARY_PATH': '/home/steam/valheim/doorstop_libs'
[ODIN][DEBUG] - Env var default 'DYLD_INSERT_LIBRARIES': '/home/steam/valheim/doorstop_libs/libdoorstop_x64.so'
[ODIN][DEBUG] - Returning environment
[ODIN][DEBUG] - Checking for BepInEx specific files...
[ODIN][DEBUG] - Path /home/steam/valheim/unstripped_corlib exists
[ODIN][DEBUG] - Path /home/steam/valheim/doorstop_libs/libdoorstop_x64.so exists
[ODIN][DEBUG] - Path /home/steam/valheim/doorstop_libs exists
[ODIN][DEBUG] - Path /home/steam/valheim/BepInEx/core/BepInEx.Preloader.dll exists
[ODIN][DEBUG] - Yay! looks like we found all the required files for BepInEx to run! <3
[ODIN][DEBUG] - BepInEx Environment:
[ODIN][DEBUG] - BepInExEnvironment {
[ODIN][DEBUG] -     ld_preload: "libdoorstop_x64.so",
[ODIN][DEBUG] -     ld_library_path: "./linux64:/home/steam/valheim/doorstop_libs",
[ODIN][DEBUG] -     doorstop_enable: "TRUE",
[ODIN][DEBUG] -     doorstop_invoke_dll: "/home/steam/valheim/BepInEx/core/BepInEx.Preloader.dll",
[ODIN][DEBUG] -     doorstop_corlib_override_path: "/home/steam/valheim/unstripped_corlib",
[ODIN][DEBUG] -     dyld_library_path: "/home/steam/valheim/doorstop_libs",
[ODIN][DEBUG] -     dyld_insert_libraries: "/home/steam/valheim/doorstop_libs/libdoorstop_x64.so",
[ODIN][DEBUG] - }
[ODIN][DEBUG] - Env var default 'WEBHOOK_URL': ''
[ODIN][DEBUG] - Skipping notification, no webhook supplied!
```


Update: In which log file do I find something about the valheimplus config?

**Reply by mbround18** [reply]
it should be grep with debug like you did, let me take a look at the code in that area :)

**Reply by nachtgold** [reply]
Could you find something out?

**Reply by mbround18** [reply]
Still diving through it, had some work to do. Should be able to get after it later today after work.

**Reply by mbround18** [reply]
@nachtgold When you run the server after modifying the config is it spitting out a file with the same name and a `.new` at the end?

**Reply by mbround18** [reply]
Alright, I have a server instance up and running. I have replaced any and all `false` in the `ValheimPlus.cfg` with `true`. Now i am going to give it a reboot and see if it persists the config.

**Reply by mbround18** [reply]
I see what you mean, it did overwrite the `valheim_plus,cfg` file but created a `.new` file for the other configs. Ill debug it.

**Reply by mbround18** [reply]
I dont think the issue exists with this container 0.o ValheimPlus itself writes its own config, when i monitor the files extracted it outputs the config files below: 

![image](https://user-images.githubusercontent.com/12646562/116456195-d28af100-a816-11eb-8a40-734e634c5a38.png)

Opening the `UnixServer.zip`, it looks like it has no `valheim_plus.cfg` included in it: 
![image](https://user-images.githubusercontent.com/12646562/116456370-0e25bb00-a817-11eb-8f62-bdd0f3e4ee21.png)

I will see about logging an issue on the ValheimPlus github

**Reply by mbround18** [reply]
@nachtgold after doing some more debugging with this, what exactly are you changing in the `valheim_plus.cfg` ? I noticed when i change values for keys inside the config file it does not modify what i have set but if i change comments or add content that is not in their originally it will overwrite it.

**Reply by nachtgold** [reply]
I commented out all variables which I don't want to use and changed `deathPenaltyMultiplier` to -100. But according to what you said, I will try it, without commenting anything.

**Reply by nachtgold** [reply]
@mbround18 Oh I made two mistakes. I only checked the last few lines of the file and there I had only commented out, no real change. But the server replaced all commented lines and keeped also my value for `deathPenaltyMultiplier` 😳 And the second mistake is, that I have to enable the block `[Player]` a few lines before, which I did not.

I'm very sorry about that. I didn't expect such magic.

**Reply by nachtgold** [reply]
PS: I also don't have a file with ending `.new` for valheim_plus.cfg

**Reply by mbround18** [reply]
@nachtgold correct, its expected that you wouldnt have a config ending in `.new` as my screenshots above in my prior comment, they don't bundle the config file in the zip but instead they write their own config.

Odin's hand off and is no longer responsible for the mod once its extracted because thats where it no longer manages anything in the servers bepinex folder :P
